src/ripple/app/consensus/RCLValidations.cpp

```diff
@@ -0,0 +1,254 @@
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/app/consensus/RCLValidations.h>
+#include <ripple/app/ledger/LedgerMaster.h>
+#include <ripple/app/main/Application.h>
+#include <ripple/app/misc/NetworkOPs.h>
+#include <ripple/app/misc/ValidatorList.h>
+#include <ripple/basics/Log.h>
+#include <ripple/basics/StringUtilities.h>
+#include <ripple/basics/chrono.h>
+#include <ripple/consensus/LedgerTiming.h>
+#include <ripple/core/DatabaseCon.h>
+#include <ripple/core/JobQueue.h>
+#include <ripple/core/TimeKeeper.h>
+#include <memory>
+#include <mutex>
+#include <thread>
+
+namespace ripple {
+
+RCLValidationsPolicy::RCLValidationsPolicy(Application& app) : app_(app)
+{
+    staleValidations_.reserve(512);
+}
+
+NetClock::time_point
+RCLValidationsPolicy::now() const
+{
+    return app_.timeKeeper().closeTime();
+}
+
+void
+RCLValidationsPolicy::onStale(RCLValidation&& v)
+{
+    // Store the newly stale validation; do not do significant work in this
+    // function since this is a callback from Validations, which may be
+    // doing other work.
+
+    ScopedLockType sl(staleLock_);
+    staleValidations_.emplace_back(std::move(v));
+    if (staleWriting_)
+        return;
+
+    staleWriting_ = true;
+    app_.getJobQueue().addJob(
+        jtWRITE, "Validations::doStaleWrite", [this](Job&) {
+            auto event =
+                app_.getJobQueue().getLoadEventAP(jtDISK, "ValidationWrite");
+            ScopedLockType sl(staleLock_);
+            doStaleWrite(sl);
+        });
+}
+
+void
+RCLValidationsPolicy::flush(hash_map<PublicKey, RCLValidation>&& remaining)
+{
+    bool anyNew = false;
+    {
+        ScopedLockType sl(staleLock_);
+
+        for (auto const& keyVal : remaining)
+        {
+            staleValidations_.emplace_back(std::move(keyVal.second));
+            anyNew = true;
+        }
+
+        // If we have new validations to write and there isn't a write in
+        // progress already, then write to the database synchronously.
+        if (anyNew && !staleWriting_)
+        {
+            staleWriting_ = true;
+            doStaleWrite(sl);
+        }
+
+        // Handle the case where flush() is called while a queuedWrite
+        // is already in progress.
```

Comment:
@wilsonianb and @bachase I think the loop is still useful.  This is a thread synchronization point.  A different thread may have called `doStaleWrite()`.  If that thread is running when we enter `flush()` at shutdown, then `staleWriting_` will be `true` (line 89).  So we'll skip calling `doStaleWrtie()` ourselves and we'll let the other thread do our job for us.  However when we exit this function all of the stale cache should be flushed to the database.  The loop makes sure we don't leave the function until the stale cache is all safely in the database.

I had hoped the comments were sufficient to describe what we're up to. but maybe not.  @wilsonianb, do you have a suggested change for the comments?